L1Trigger/TrackFindingTMTT/interface/Histos.h

```diff
@@ -0,0 +1,661 @@
+#ifndef L1Trigger_TrackFindingTMTT_Histos_h
+#define L1Trigger_TrackFindingTMTT_Histos_h
+
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "CommonTools/UtilAlgos/interface/TFileService.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Settings.h"
+#include "L1Trigger/TrackFindingTMTT/interface/L1track3D.h"
+#include "L1Trigger/TrackFindingTMTT/interface/TrackerGeometryInfo.h"
+
+#include "boost/numeric/ublas/matrix.hpp"
+using boost::numeric::ublas::matrix;
+
+#include <vector>
+#include <map>
+#include <string>
+
+using namespace std;
+
+class TH1F;
```

Comment:
We're unsure of the best solution to this. The Histo class is not just validation code (to identify bugs), but is also used to study & optimise the L1 tracking performance. One option would be to disable it with a pragma in the release, so allowing edm::stream to be used for the EDProducer, but allowing it to be reenabled for L1 track studies. An alternative would be to make EDProducts of all the histogrammed quantities, but this would mean adding a lot of objects, purely for use by Histos, to the DataFormats package, and since Histos uses MC truth & association too, this would become very painful if we were required to have separate EDProducts for reco, truth & assocation. I wonder if we can somehow submit an empty version of the Histos class to the release, in a way that would allow the full version to be linked with the code for development work? But I've not found a clever way to do this.

Reply:
A possibility could be to use the DQM to book and manage all the histograms.
Would that work here ?

Reply:
The offline tracking group manages to optimize their performance without injecting histograms into production code. They use the MultiTrackValidator that runs in the Validation step (and lives in that subsystem). I encourage adopting a similar approach.

Reply:
I guess you should open an issue in GitHub and add this rewrite as a target for CMSSW 12.0.

Unless it actually causes problems with running the jobs, I wouldn't consider this a blocker for 11.1.

Reply:
As mentioned elsewhere, moving the histograms out of the EDProducer to a separate EDAnalyzer is something we will do for a planned major rewrite of the code. But it would be very effort intensive to do it now, and also clutter up DataFormats with lots of EDProducts used purely for these histograms. Therefore for the moment, we are using edm::stream, with the histograms booked in the GlobalCache, and filled later protected by a mutex lock, such that only one thread can fill them at a time.

Reply:
Let me add that as a general rule `edm::one` is preferred over `edm::stream`+mutexes, because with `edm::one` the framework can schedule work such that the "at most one running" condition is satisfied without blocking a thread. (and then there are a bunch of a second-order subtleties that can affect the optimal choice)

Said that, for this particular case  the `edm::stream` + mutexes is probably good enough (for 11_1), assuming that the protected critical sections are not "too long".